### PR TITLE
 UFAL/Enhanced type-bind feature

### PR DESCRIPTION
**Pull request by milanmajchrak**
| Phases            | MP | MM  |  MB  | MR |   JM | Total  |
|-----------------|----:|----:|----:|-----:|-----:|-------:|
| ETA                  |  0  |  0  |    0 |     0 |      0 |        0 |
| Developing      |  0  |  0  |    0 |    0 |      0 |         0 |
| Review             |  0  |  0  |    0 |    0 |      0 |         0 |
| Total                |   -  |   -  |   -   |  -    |   -    |         0 |
| ETA est.             |      |      |       |       |         |         0 |
| ETA cust.           |   -  |   -  |   -  |   -   |   -     |        0 |

### CODE REVIEW

File: dspace-api/src/main/java/org/dspace/app/util/DCInput.java [hunk 1]
```diff
@@ -559,6 +568,21 @@ public boolean isAllowedFor(String typeName) {
...
 
+    /**
+     * Decides if this field is valid for the document type
+     * Check if one of the typeName is in the typeBind list
+     *
+     * @param typeNames List of document type names e.g. ["VIDEO"]
+     * @return true when there is no type restriction or typeName is allowed
+     */
+    public boolean isAllowedFor(List<String> typeNames) {
+        if (typeBind.size() == 0) {
```
Comment (vidiecan) [inline dspace-api/src/main/java/org/dspace/app/util/DCInput.java:579]: isEmpty

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/submit/step/validation/MetadataValidation.java [hunk 2]
```diff
@@ -68,22 +68,46 @@ public List<ErrorRest> validate(SubmissionService submissionService, InProgressS
...
+        // Get the list of type-bind fields. It could be in the form of schema.element.qualifier=>metadata_field, or
+        // just metadata_field
+        List<String> typeBindFields = Arrays.asList(
+                configurationService.getArrayProperty("submit.type-bind.field", new String[0]));
+
+        for (String typeBindField : typeBindFields) {
+            String typeBFKey = typeBindField;
+            // If the type-bind field is in the form of schema.element.qualifier=>metadata_field, split it and get the
+            // metadata field
+            if (typeBindField.contains("=>")) {
```
Comment (vidiecan) [inline dspace-server-webapp/src/main/java/org/dspace/app/rest/submit/step/validation/MetadataValidation.java:82]: id `=>` somewhere specified, would be great to have it as a constant or similar

**Review by vidiecan — CHANGES_REQUESTED** [review]

File: dspace-api/src/main/java/org/dspace/app/util/DCInput.java [hunk 3]
```diff
@@ -273,6 +273,15 @@ public DCInput(Map<String, String> fieldMap, Map<String, List<String>> listMap,
 
     }
 
+    private void insertToTypeBind(String typeBindDef) {
+        if (typeBindDef != null && typeBindDef.trim().length() > 0) {
```
Comment (vidiecan) [inline dspace-api/src/main/java/org/dspace/app/util/DCInput.java:277]: typeBindDef.trim().isEmpty()
@milanmajchrak TT

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/submit/step/validation/MetadataValidation.java [hunk 4]
```diff
@@ -167,7 +193,11 @@ public List<ErrorRest> validate(SubmissionService submissionService, InProgressS
                     }
                 }
             }
+
         }
+
+
+        // Begin the actual validation loop
```
Comment (vidiecan) [inline dspace-server-webapp/src/main/java/org/dspace/app/rest/submit/step/validation/MetadataValidation.java:200]: ?

**Review by vidiecan — CHANGES_REQUESTED** [review]

File: dspace-api/src/main/java/org/dspace/app/util/DCInput.java (hunk 3 above)
Comment (milanmajchrak) [reply dspace-api/src/main/java/org/dspace/app/util/DCInput.java:277]: changed

**Review by vidiecan — APPROVED** [review]

**Review by vidiecan — APPROVED** [review]

**Pull request merged**